### PR TITLE
cmake: fix relative path calculate error

### LINKED ISSUE #74647: After menuconfig, west build failed.
If you modify the configuration through menuconfig, an error will be reported when compiling again.

For example:

1. west build -b bbc_microbit_v2 zephyr/tests/bluetooth/shell/ -p auto  
2. west build -t menuconfig, type `/` set `BT_DEBUG_MONITOR_UART` to `y` and enter save.
3. west build

```
modules/hal/nordic/nrfx/mdk -I/home/mi/works/github/zephyrproject/zephyr/modules/hal_nordic/nrfx/. -I/media/mi/5e58d875-5e34-6b43-8936-d956a2f60c04/works/github/zephyrproject/modules/hal/st/sensor/stmemsc/lis2mdl_STdC/driver/ -I/media/mi/5e58d875-5e34-6b43-8936-d956a2f60c04/works/github/zephyrproject/modules/crypto/tinycrypt/lib/include -I/home/mi/works/github/zephyrproject/zephyr/include/zephyr/arch/arm/cortex_m/scripts -D__GCC_LINKER_CMD__ -DLINKER_ZEPHYR_PREBUILT -E /home/mi/works/github/zephyrproject/zephyr/include/zephyr/arch/arm/cortex_m/scripts/linker.ld -P -o linker_zephyr_pre0.cmd && /usr/bin/cmake -E cmake_transform_depfile Ninja gccdepfile /home/mi/works/github/zephyrproject/zephyr/tests/bluetooth/shell /home/mi/works/github/zephyrproject/zephyr /home/mi/works/github/zephyrproject/build /home/mi/works/github/zephyrproject/build/zephyr /home/mi/works/github/zephyrproject/build/zephyr/linker_zephyr_pre0.cmd.dep /home/mi/works/github/zephyrproject/build/CMakeFiles/d/af1ffabef38d460b6c582d0d863c22e37f606b7bb6feae921eb21593aefd1607.d
In file included from /home/mi/works/github/zephyrproject/zephyr/include/zephyr/arch/arm/cortex_m/scripts/linker.ld:145:
/home/mi/works/github/zephyrproject/build/zephyr/include/generated/snippets-rom-start.ld:4:38: fatal error: ../../../../../../../media/mi/5e58d875-5e34-6b43-8936-d956a2f60c04/works/github/zephyrproject/zephyr/arch/arm/core/vector_table.ld: No such file or directory
    4 | /* Sort key: "0x0vectors" */#include "../../../../../../../media/mi/5e58d875-5e34-6b43-8936-d956a2f60c04/works/github/zephyrproject/zephyr/arch/arm/core/vector_table.ld"
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[23/237] Building C object zephyr/lib/libc/common/CMakeFiles/lib__libc__common.dir/source/stdlib/malloc.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /media/mi/5e58d875-5e34-6b43-8936-d956a2f60c04/works/github/zephyrproject/build
```

### PR DESCRIPTION
**Pull request by LingaoM**
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/74647

image

### CODE REVIEW

**Review by ithinuel — CHANGES_REQUESTED** [review]
This removes the use of relative path which was intentionally added in https://github.com/zephyrproject-rtos/zephyr/pull/16368.

cc @marc-hb is it still required ?

**Comment by marc-hb** [comment]
> The build directory absolutely is local generated directory, not public for other.

No, the build directory is not "100% private": to make sure the build is reproducible you need to compare object files.

Reproducible builds are important for security, for instance they help with "JiaT75" issues. Comparing build directories also helps catching small but harmful differences between CI and developers and between developers.
 
For more context see #50205, #14593, https://www.cisa.gov/sbom, https://reproducible-builds.org/ etc.

**Review by marc-hb — DISMISSED** [review]
> This removes the use of relative path which was intentionally added in https://github.com/zephyrproject-rtos/zephyr/pull/16368.

This is indeed a logical revert of d1b4da96e1484a276c21c3, so at least the commit message is wrong. It currently says "fix relative path calculate error" but this commit does not fix any "calculate error", instead it reverts to an absolute, non reproducible path.

Thanks @ithinuel for the good catch.

**Comment by LingaoM** [comment]
@marc-hb @ithinuel I've reworked a version that will still use relative paths, but adds support for link paths.

File: cmake/modules/extensions.cmake [hunk 1]
```diff
@@ -1335,7 +1335,10 @@ function(zephyr_linker_sources location)
     endif()
 
     # Find the relative path to the linker file from the include folder.
-    file(RELATIVE_PATH relpath ${ZEPHYR_BASE}/include ${path})
+    get_filename_component(ZEPHYR_BASE_REALPATH "${ZEPHYR_BASE}/include" REALPATH)
+    get_filename_component(PATH_REALPATH "${path}" REALPATH)
```
Comment (tejlmand) [inline cmake/modules/extensions.cmake:1339]: The command `get_filename_component(... REALPATH)` has been superseded by `file(REAL_PATH ...)` cmake.
> `get_filename_component`
> 
> Get a specific component of a full filename.
>
> Changed in version 3.20: This command has been superseded by the [cmake_path()](https://cmake.org/cmake/help/latest/command/cmake_path.html#command:cmake_path) command, except for REALPATH, which is now offered by [file(REAL_PATH)](https://cmake.org/cmake/help/latest/command/file.html#real-path),

https://cmake.org/cmake/help/latest/command/get_filename_component.html#command:cmake_path

but please note that `file(REAL_PATH ...)` is affected by [CMP0152](https://cmake.org/cmake/help/latest/policy/CMP0152.html)., and before you ask, then yes, `get_filename_component(... REAL_PATH)` suffers the same issue related to symlink resolving, but doesn't have a policy for controlling this (always fixed at OLD behavior).

Which again shows that the use of symlinks can be tricky.

Also, variables intended for local use should be written in lowercase, like `path_realpath` and `zephyr_base_realpath`.
And please use another name than `zephyr_base_realpath`, as it is not the real path of Zephyr you are defining, but a subfolder (`include`).

**Review by tejlmand — CHANGES_REQUESTED** [review]

**Pull request closed without merging**